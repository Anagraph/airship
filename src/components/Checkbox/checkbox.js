import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled, { css, keyframes } from 'styled-components';
import { rgba, darken } from 'polished';
import { theme } from '../../constants';

const stroke = keyframes`
  100% {
    stroke-dashoffset: 0;
  }
`;

const Wrapper = styled.label`
  display: inline-flex;
  align-items: center;
`;

const StyledCheckbox = styled.div`
  position: relative;
`;

const Label = styled.div`
  margin-left: 4px;
  font: 400 12px/20px 'Roboto';
`;

const Svg = styled.svg`
  pointer-events: none;
  width: 16px;
  height: 16px;
  overflow: hidden;
  border-radius: 3px;
  position: relative;
  display: block;
  box-sizing: border-box;
`;

const Check = styled.polyline`
  fill: none;
  stroke-width: 2;
  stroke: ${props => props.theme.brand01};
  transform-origin: 50% 50%;
  stroke-dasharray: 48;
  stroke-dashoffset: 48;
  transform: translate(2px, 4px);
`;
Check.defaultProps = {
  theme,
};

const CheckboxInput = styled.input`
  appearance: none;
  background: none;
  border-radius: 3px;
  border: 1px solid ${props => props.theme.ui04};
  cursor: pointer;
  height: 16px;
  left: 0;
  margin: 0;
  padding: 0;
  position: absolute;
  top: 0;
  transition: border 0.2s ease;
  width: 16px;

  &:focus {
    outline: none;
    border: 2px solid ${props => props.theme.brand01};
  }

  &:hover {
    border-color: ${props => props.theme.brand03};
  }

  &:checked {
    border-color: ${props => props.theme.brand01};

    & + ${Svg} ${Check} {
      animation: ${stroke} 0.3s cubic-bezier(0.65, 0, 0.45, 1) 300ms forwards;
    }
  }

  ${props => props.error && css`
    background: ${props => rgba(props.theme.support01, 0.12)};
    border-color: ${props => props.theme.support01};

    &:focus {
      border-color: ${props => props.theme.support01};
    }

    &:hover {
      border-color: ${props => darken(0.16, props.theme.support01)};
    }

    &:checked {
      border-color: ${props => props.theme.support01};

      & + ${Svg} ${Check} {
       stroke: ${props => props.theme.support01};
      }
    }
  `};

  &:disabled {
    background-color: ${props => props.theme.ui02};
    cursor: not-allowed;
    border-color: ${props => props.theme.ui04};

    &:checked + ${Svg} ${Check} {
      stroke: ${props => props.theme.ui04};
    }
  }
`;
CheckboxInput.defaultProps = {
  theme,
};

class Checkbox extends Component {
  state = {
    checked: !!this.props.checked,
  };

  componentWillReceiveProps(nextProps) {
    if (this.state.checked !== nextProps.checked) {
      this.setState({ checked: nextProps.checked });
    }
  }

  clickHandler = () => {
    const { onChange, disabled } = this.props;

    if (disabled) return;

    this.setState(
      state => ({ ...state, checked: !state.checked }),
      () => onChange && onChange(this.state)
    );
  };

  render() {
    const { as, children, disabled, error, htmlFor, name } = this.props;
    const { checked } = this.state;
    const WrapperComponent = as !== 'div' ? Wrapper.withComponent(as) : Wrapper;

    return (
      <WrapperComponent htmlFor={htmlFor}>
        <StyledCheckbox>
          <CheckboxInput
            id={htmlFor}
            type="checkbox"
            name={name}
            onChange={this.clickHandler}
            disabled={disabled}
            checked={checked}
            error={error}
          />
          <Svg>
            <Check points="1.65093994 3.80255127 4.48919678 6.97192383 10.3794556 0.717346191" />
          </Svg>
        </StyledCheckbox>

        {children && <Label>{children}</Label>}
      </WrapperComponent>
    );
  }
}

Checkbox.defaultProps = {
  as: 'div',
  checked: false,
  disabled: false,
  error: false,
  name: '',
  onChange: () => {},
};

Checkbox.propTypes = {
  as: PropTypes.string,
  checked: PropTypes.bool,
  children: PropTypes.node,
  disabled: PropTypes.bool,
  error: PropTypes.bool,
  htmlFor: PropTypes.string.isRequired,
  name: PropTypes.string,
  onChange: PropTypes.func,
};

export default Checkbox;
