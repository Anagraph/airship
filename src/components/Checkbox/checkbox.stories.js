import React from 'react';
import { storiesOf } from '@storybook/react';
import { ThemeProvider } from 'styled-components';
import { action } from '@storybook/addon-actions';
import Checkbox from './checkbox';
import { theme } from '../../constants';

const CUSTOM_THEME = {
  ...theme,
  brand01: '#FABADA',
};

storiesOf('Checkbox', module)
  .add('Default', () => (
    <div>
      <h3 className="header">Default</h3>
      <Checkbox htmlFor="default" />

      <h3 className="header">With label</h3>
      <Checkbox htmlFor="with-label">
        Hello there
      </Checkbox>

      <h3 className="header">Checked</h3>
      <Checkbox htmlFor="checked" checked>
        Hello there
      </Checkbox>

      <h3 className="header">With error</h3>
      <Checkbox htmlFor="error" error>
        Hello there
      </Checkbox>

      <h3 className="header">Checked with error</h3>
      <Checkbox htmlFor="checked-error" checked error>
        Hello there
      </Checkbox>
    </div>
  ))
  .add('Disabled', () => (
    <div>
      <h3 className="header">Default</h3>
      <Checkbox htmlFor="disabled" disabled />

      <h3 className="header">With label</h3>
      <Checkbox htmlFor="with-label" disabled>
        Hello there
      </Checkbox>

      <h3 className="header">Checked</h3>
      <Checkbox htmlFor="checked" disabled checked>
        Hello there
      </Checkbox>

      <h3 className="header">With error</h3>
      <Checkbox htmlFor="checked" disabled error>
        Hello there
      </Checkbox>

      <h3 className="header">With error checked</h3>
      <Checkbox htmlFor="checked" disabled error checked>
        Hello there
      </Checkbox>
    </div>
  ))
  .add('With event', () => (
    <div>
      <h3 className="header">Default</h3>
      <Checkbox htmlFor="default" onChange={action('onChange')}>
        Click me!
      </Checkbox>

      <h3 className="header">Checked</h3>
      <Checkbox htmlFor="checked" onChange={action('onChange')} checked>
        Click me!
      </Checkbox>

      <h3 className="header">Disabled</h3>
      <Checkbox htmlFor="disabled" onChange={action('onChange')} disabled>
        Click me!
      </Checkbox>
    </div>
  ))
  .add('With custom theme', () => (
    <ThemeProvider theme={CUSTOM_THEME}>
      <div>
        <h3 className="header">Default</h3>
        <Checkbox htmlFor="default">
          Click me!
        </Checkbox>

        <h3 className="header">Checked</h3>
        <Checkbox htmlFor="checked" checked>
          Click me!
        </Checkbox>

        <h3 className="header">Disabled</h3>
        <Checkbox htmlFor="disabled" disabled>
          Click me!
        </Checkbox>
      </div>
    </ThemeProvider>
  ));
