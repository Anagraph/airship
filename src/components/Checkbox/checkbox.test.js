import React from 'react';
import Checkbox from './checkbox';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

describe('render', () => {
  it('renders without crashing', () => {
    const component = renderer.create(<Checkbox value="wadus" />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with label', () => {
    const component = renderer.create(
      <Checkbox value="wadus" checked>
        Hola
      </Checkbox>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('onChange', () => {
    const spy = jest.fn();
    const component = mount(
      <Checkbox value="wadus" onChange={spy}>
        Hola
      </Checkbox>
    );

    component.find('input').simulate('change');
    expect(spy).toHaveBeenCalled();
  });
});
