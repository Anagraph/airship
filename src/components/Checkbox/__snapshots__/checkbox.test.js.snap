// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render renders with label 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  position: relative;
}

.c7 {
  margin-left: 4px;
  font: 400 12px/20px 'Roboto';
}

.c4 {
  pointer-events: none;
  width: 16px;
  height: 16px;
  overflow: hidden;
  border-radius: 3px;
  position: relative;
  display: block;
  box-sizing: border-box;
}

.c6 {
  fill: none;
  stroke-width: 2;
  stroke: #1785FB;
  -webkit-transform-origin: 50% 50%;
  -ms-transform-origin: 50% 50%;
  transform-origin: 50% 50%;
  stroke-dasharray: 48;
  stroke-dashoffset: 48;
  -webkit-transform: translate(2px,4px);
  -ms-transform: translate(2px,4px);
  transform: translate(2px,4px);
}

.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: none;
  border-radius: 3px;
  border: 1px solid #D1D5D7;
  cursor: pointer;
  height: 16px;
  left: 0;
  margin: 0;
  padding: 0;
  position: absolute;
  top: 0;
  -webkit-transition: border 0.2s ease;
  transition: border 0.2s ease;
  width: 16px;
}

.c2:focus {
  outline: none;
  border: 2px solid #1785FB;
}

.c2:hover {
  border-color: #47DB99;
}

.c2:checked {
  border-color: #1785FB;
}

.c2:checked + .c3 .c5 {
  -webkit-animation: gGaDtf 0.3s cubic-bezier(0.65,0,0.45,1) 300ms forwards;
  animation: gGaDtf 0.3s cubic-bezier(0.65,0,0.45,1) 300ms forwards;
}

.c2:disabled {
  background-color: #F5F5F5;
  cursor: not-allowed;
  border-color: #D1D5D7;
}

.c2:disabled:checked + .c3 .c5 {
  stroke: #D1D5D7;
}

<label
  className="c0"
  htmlFor="test"
>
  <div
    className="c1"
  >
    <input
      checked={true}
      className="c2"
      disabled={false}
      id="test"
      name=""
      onChange={[Function]}
      type="checkbox"
    />
    <svg
      className="c3 c4"
    >
      <polyline
        className="c5 c6"
        points="1.65093994 3.80255127 4.48919678 6.97192383 10.3794556 0.717346191"
      />
    </svg>
  </div>
  <div
    className="c7"
  >
    Hola
  </div>
</label>
`;

exports[`render renders without crashing 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  position: relative;
}

.c4 {
  pointer-events: none;
  width: 16px;
  height: 16px;
  overflow: hidden;
  border-radius: 3px;
  position: relative;
  display: block;
  box-sizing: border-box;
}

.c6 {
  fill: none;
  stroke-width: 2;
  stroke: #1785FB;
  -webkit-transform-origin: 50% 50%;
  -ms-transform-origin: 50% 50%;
  transform-origin: 50% 50%;
  stroke-dasharray: 48;
  stroke-dashoffset: 48;
  -webkit-transform: translate(2px,4px);
  -ms-transform: translate(2px,4px);
  transform: translate(2px,4px);
}

.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: none;
  border-radius: 3px;
  border: 1px solid #D1D5D7;
  cursor: pointer;
  height: 16px;
  left: 0;
  margin: 0;
  padding: 0;
  position: absolute;
  top: 0;
  -webkit-transition: border 0.2s ease;
  transition: border 0.2s ease;
  width: 16px;
}

.c2:focus {
  outline: none;
  border: 2px solid #1785FB;
}

.c2:hover {
  border-color: #47DB99;
}

.c2:checked {
  border-color: #1785FB;
}

.c2:checked + .c3 .c5 {
  -webkit-animation: gGaDtf 0.3s cubic-bezier(0.65,0,0.45,1) 300ms forwards;
  animation: gGaDtf 0.3s cubic-bezier(0.65,0,0.45,1) 300ms forwards;
}

.c2:disabled {
  background-color: #F5F5F5;
  cursor: not-allowed;
  border-color: #D1D5D7;
}

.c2:disabled:checked + .c3 .c5 {
  stroke: #D1D5D7;
}

<label
  className="c0"
  htmlFor="test"
>
  <div
    className="c1"
  >
    <input
      checked={false}
      className="c2"
      disabled={false}
      id="test"
      name=""
      onChange={[Function]}
      type="checkbox"
    />
    <svg
      className="c3 c4"
    >
      <polyline
        className="c5 c6"
        points="1.65093994 3.80255127 4.48919678 6.97192383 10.3794556 0.717346191"
      />
    </svg>
  </div>
</label>
`;

exports[`render uses htmlFor for the label, and adds it as the input id 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  position: relative;
}

.c4 {
  pointer-events: none;
  width: 16px;
  height: 16px;
  overflow: hidden;
  border-radius: 3px;
  position: relative;
  display: block;
  box-sizing: border-box;
}

.c6 {
  fill: none;
  stroke-width: 2;
  stroke: #1785FB;
  -webkit-transform-origin: 50% 50%;
  -ms-transform-origin: 50% 50%;
  transform-origin: 50% 50%;
  stroke-dasharray: 48;
  stroke-dashoffset: 48;
  -webkit-transform: translate(2px,4px);
  -ms-transform: translate(2px,4px);
  transform: translate(2px,4px);
}

.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: none;
  border-radius: 3px;
  border: 1px solid #D1D5D7;
  cursor: pointer;
  height: 16px;
  left: 0;
  margin: 0;
  padding: 0;
  position: absolute;
  top: 0;
  -webkit-transition: border 0.2s ease;
  transition: border 0.2s ease;
  width: 16px;
}

.c2:focus {
  outline: none;
  border: 2px solid #1785FB;
}

.c2:hover {
  border-color: #47DB99;
}

.c2:checked {
  border-color: #1785FB;
}

.c2:checked + .c3 .c5 {
  -webkit-animation: gGaDtf 0.3s cubic-bezier(0.65,0,0.45,1) 300ms forwards;
  animation: gGaDtf 0.3s cubic-bezier(0.65,0,0.45,1) 300ms forwards;
}

.c2:disabled {
  background-color: #F5F5F5;
  cursor: not-allowed;
  border-color: #D1D5D7;
}

.c2:disabled:checked + .c3 .c5 {
  stroke: #D1D5D7;
}

<label
  className="c0"
  htmlFor="test"
>
  <div
    className="c1"
  >
    <input
      checked={false}
      className="c2"
      disabled={false}
      id="test"
      name=""
      onChange={[Function]}
      type="checkbox"
    />
    <svg
      className="c3 c4"
    >
      <polyline
        className="c5 c6"
        points="1.65093994 3.80255127 4.48919678 6.97192383 10.3794556 0.717346191"
      />
    </svg>
  </div>
</label>
`;
